File: 9. Convert infix expression to postfix using a stack.cpp
/*
 Task: Convert infix expression to postfix using a stack
*/

#include <bits/stdc++.h>
using namespace std;

int prec(char c)
{
    if (c == '+' || c == '-')
        return 1;
    if (c == '*' || c == '/')
        return 2;
    return 0;
}

string infixToPostfix(const string& exp)
{
    stack<char> st;
    string out;
    for (char c : exp)
    {
        if (isalnum(c))
            out.push_back(c);
        else if (c == '(')
            st.push(c);
        else if (c == ')')
        {
            while (!st.empty() && st.top() != '(')
            {
                out.push_back(st.top());
                st.pop();
            }
            if (!st.empty())
                st.pop();
        }
        else if (isspace(c))
            continue;
        else
        {
            while (!st.empty() && prec(st.top()) >= prec(c))
            {
                out.push_back(st.top());
                st.pop();
            }
            st.push(c);
        }
    }
    while (!st.empty())
    {
        out.push_back(st.top());
        st.pop();
    }
    return out;
}

int main()
{
    string s = "a+b*(c^d-e)^(f+g*h)-i";
    cout << infixToPostfix(s) << "\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 9. Convert infix expression to postfix using a stack.py
"""
Task: Convert infix expression to postfix using a stack

Time Complexity: O(n)
Space Complexity: O(n)
"""

def prec(c: str) -> int:
    if c in '+-': return 1
    if c in '*/': return 2
    return 0


def infix_to_postfix(exp: str) -> str:
    st = []
    out = []
    for c in exp:
        if c.isalnum():
            out.append(c)
        elif c == '(':
            st.append(c)
        elif c == ')':
            while st and st[-1] != '(':
                out.append(st.pop())
            if st: st.pop()
        elif c.isspace():
            continue
        else:
            while st and prec(st[-1]) >= prec(c):
                out.append(st.pop())
            st.append(c)
    while st:
        out.append(st.pop())
    return ''.join(out)

if __name__ == '__main__':
    s = 'a+b*(c^d-e)^(f+g*h)-i'
    print(infix_to_postfix(s))
